### PR TITLE
Fix CORS origin in CSV format

### LINKED ISSUE #12259: Cross-Origin Request Blocked with CSV value in CORS_ORIGIN setting
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues?q=is%3Aissue) that already describes my problem.

### Describe the Bug

Since `9.7.0` CORS has been disabled by default, I need it though so I thought might as well re-enable it but this time restrict to only specific domains. The documentation says using a CSV of domains is permitted: https://docs.directus.io/configuration/config-options/#cors

Turns out it doesn't work, because it just sets the cors origin header to the CSV value instead of matching domains and then only setting the matched domain as the origin value. The CORS origin header can only contain one domain and it needs to be the same one as the one you are on, otherwise CORS requests are invalid. See: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin

> `<origin>` Specifies an origin. Only a single origin can be specified. If the server supports clients from multiple origins, it must return the origin for the specific client making the request.

### To Reproduce

Set `CORS_ORIGIN` in `.env` file to a CSV of domains.

### Errors Shown

Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://www.gastrodino.de/api/items/products (Reason: CORS header ‘Access-Control-Allow-Origin’ does not match **‘http://localhost:4200,https://www.gastrodino.com’**

### What version of Directus are you using?

9.7.0

### What version of Node.js are you using?

16.14.2

### What database are you using?

MariaDB 10.3.34

### What browser are you using?

Chrome Latest

### What operating system are you using?

Windows 10

### How are you deploying Directus?

Ubuntu Server 20.04, npm installation with a systemd service running it

### PR DESCRIPTION
**Pull request by licitdev**
Closes #12259.

### CODE REVIEW

**Comment by rijkvanzanten** [comment]
@licitdev Express' cors module should already support this by default:

> Array - set origin to an array of valid origins. Each origin can be a String or a RegExp. For example ["http://example1.com", /\.example2\.com$/] will accept any request from “http://example1.com” or from a subdomain of “example2.com”.

I think the problem is that the CSV isn't cast as an array properly in the env reading, causing it to be recognized as a string, rather than an array of strings

**Comment by licitdev** [comment]
@rijkvanzanten I've tested it by passing a plain array, it didn't work as expected. When provided with an array, if not in the list, the origin is set to '*'. Thus I've used the `toArray()` util for parsing, similar to other env variables that allows CSV input.

This current PR sets the origin as...
'*' when `CORS_ORIGIN` is not set.
'matching origin' when matches any item in `CORS_ORIGIN`
'first origin' when no match is found

**Comment by rijkvanzanten** [comment]
> @rijkvanzanten I've tested it by passing a plain array, it didn't work as expected. When provided with an array, if not in the list, the origin is set to '*'. Thus I've used the `toArray()` util for parsing, similar to other env variables that allows CSV input.
> 
> This current PR sets the origin as... '*' when `CORS_ORIGIN` is not set. 'matching origin' when matches any item in `CORS_ORIGIN` 'first origin' when no match is found

Odd! Thanks for confirming that 👍🏻

**Review by rijkvanzanten — CHANGES_REQUESTED** [review]

File: api/src/middleware/cors.ts [hunk 1]
```diff
@@ -1,12 +1,23 @@
...
 let corsMiddleware: RequestHandler = (req, res, next) => next();
 
 if (env.CORS_ENABLED === true) {
+	const originWhitelist = env.CORS_ORIGIN ? toArray(env.CORS_ORIGIN) : false;
+
 	corsMiddleware = cors({
-		origin: env.CORS_ORIGIN || true,
+		origin: function (origin, callback) {
+			if (!originWhitelist) {
+				callback(null, true);
```
[the remark this answers is not in the document]
Comment (licitdev) [reply api/src/middleware/cors.ts:14]: In my tests, when setting to `true`, the origin will be set as `*`.
Will change it to increase clarity. Thanks!

[the remark this answers is not in the document]
Comment (rijkvanzanten) [reply api/src/middleware/cors.ts]: Huh! Are the express docs misleading then, or am I missing something here? 🤔 

https://expressjs.com/en/resources/middleware/cors.html#configuration-options

> The origin argument to the callback can be any value allowed for the origin option of the middleware, except a function. See the [configuration options](https://expressjs.com/en/resources/middleware/cors.html#configuration-options) section for more information on all the possible value types.

> Configuration Options
origin: Configures the Access-Control-Allow-Origin CORS header. Possible values:
>  - Boolean - set origin to true to reflect the [request origin](http://tools.ietf.org/html/draft-abarth-origin-09), as defined by req.header('Origin'), or set it to false to disable CORS.
>  - String - set origin to a specific origin. For example if you set it to "http://example.com" only requests from “http://example.com” will be allowed.

[the remark this answers is not in the document]
Comment (rijkvanzanten) [reply api/src/middleware/cors.ts]: Maybe express defaults to `*` when the Origin header is missing? Did you test it from a browser or from an API tool?

[the remark this answers is not in the document]
Comment (licitdev) [reply api/src/middleware/cors.ts]: @rijkvanzanten I am not sure why. Tested in the admin app, [test-cors.org](https://www.test-cors.org/) and also through an API tool (without origin). 
Screenshot 2022-03-21 at 10 02 13 PM
Screenshot 2022-03-21 at 10 02 29 PM

[the remark this answers is not in the document]
Comment (rijkvanzanten) [reply api/src/middleware/cors.ts]: It does seem to work as expected on my end. The Origin you're requesting doesn't exist in the `origin` configuration you used in the last screenshot, so the header isn't returned. When I configure `origin: ['http://localhost:8055', 'https://rijks.website'],` and set the Origin header to `https://rijks.website`, I'm correctly getting `Access-Control-Allow-Origin: https://rijks.website` returned

[the remark this answers is not in the document]
Comment (licitdev) [reply api/src/middleware/cors.ts]: But if you access from a third party website, a `*` gets returned. Which should not be the case as we are limiting the origins to be from our whitelist only?

[the remark this answers is not in the document]
Comment (licitdev) [reply api/src/middleware/cors.ts]: Note: I was testing through the Vite development server on `:8080`, thus encountered the above issue.

**Comment by rijkvanzanten** [comment]
Allllllll this confusion stemmed from the fact that we were testing against the _dev proxy_ provided by Vite, which injects some custom magic in the CORS header. Besides the auto-casting of array like values in the .env (https://github.com/directus/directus/pull/12330), everything is actually working as expected

**Pull request closed without merging**